### PR TITLE
[SPARK-13951][ML][PYTHON] Nested Pipeline persistence

### PR DESCRIPTION
**Pull request by jkbradley**
## What changes were proposed in this pull request?

Adds support for saving and loading nested ML Pipelines from Python.  Pipeline and PipelineModel do not extend JavaWrapper, but they are able to utilize the JavaMLWriter, JavaMLReader implementations.

Also:
- Separates out interfaces from Java wrapper implementations for MLWritable, MLReadable, MLWriter, MLReader.
- Moves methods _stages_java2py, _stages_py2java into Pipeline, PipelineModel as _transfer_stage_from_java, _transfer_stage_to_java
## How was this patch tested?

Added new unit test for nested Pipelines.  Abstracted validity check into a helper method for the 2 unit tests.

### CODE REVIEW

File: python/pyspark/ml/pipeline.py [hunk 1]
```diff
@@ -206,49 +167,65 @@ def save(self, path):
...
-        """Returns an JavaMLReader instance for this class."""
+        """Returns an MLReader instance for this class."""
         return PipelineMLReader(cls)
 
     @classmethod
-    @since("2.0.0")
-    def load(cls, path):
-        """Reads an ML instance from the input path, a shortcut of `read().load(path)`."""
-        return cls.read().load(path)
+    def _transfer_stage_from_java(cls, java_stage):
```
Comment (mengxr) [inline python/pyspark/ml/pipeline.py]: It is a little confusing to call it `java_stage` while the expected input is a pipeline. Shall we rename it to `_from_java_pipeline` and `_transfer_stage_to_java` to `_to_java_pipeline`? If we have to use the same method name, I would recommend `_from_java` and `_to_java` instead.

File: python/pyspark/ml/pipeline.py [hunk 2]
```diff
@@ -206,49 +167,65 @@ def save(self, path):
...
+        py_stage = cls()
+        # Load information from java_stage to the instance.
+        py_stages = [JavaWrapper._transfer_stage_from_java(s) for s in java_stage.getStages()]
+        py_stage.setStages(py_stages)
+        py_stage._resetUid(java_stage.uid())
+        return py_stage
+
+    def _transfer_stage_to_java(self):
+        """
+        Transfer this instance to a Java object.  Used for ML persistence.
```
Comment (mengxr) [inline python/pyspark/ml/pipeline.py]: `Java object` -> `Java pipeline`

**Comment by mengxr** [comment]
Not part of this PR, in PySpark `DataFrame.write` is a property instead of a method. Shall we follow the same convention?

File: python/pyspark/ml/pipeline.py (hunk 1 above)
Comment (jkbradley) [inline python/pyspark/ml/pipeline.py]: I'd like to use the same method name, so I'll switch to ```_from/to_java```.

File: python/pyspark/ml/pipeline.py (hunk 2 above)
Comment (jkbradley) [inline python/pyspark/ml/pipeline.py]: Done

**Comment by jkbradley** [comment]
Making MLWritable.write a property sounds good.  I'll create a JIRA for it.

**Comment by mengxr** [comment]
LGTM. Merged into master. Thanks!

File: python/pyspark/ml/util.py [hunk 3]
```diff
@@ -74,18 +74,38 @@ def _randomUID(cls):
...
+
+    def context(self, sqlContext):
+        """Sets the SQL context to use for saving."""
+        raise NotImplementedError("MLWriter is not yet implemented for type: %s" % type(self))
+
+
+@inherit_doc
+class JavaMLWriter(MLWriter):
+    """
+    (Private) Specialization of :py:class:`MLWriter` for :py:class:`JavaWrapper` types
```
Comment (zhengruifeng) [inline python/pyspark/ml/util.py]: @jkbradley  @mengxr  Excuse me, why we need to add a Private annotation here and in other places?

**Pull request closed without merging**